conn.go

```diff
@@ -702,12 +703,20 @@ func (c *Conn) handleIncomingPacket(ctx context.Context, buf []byte, enqueue boo
 	} else if isHandshake {
 		markPacketAsValid()
 		for out, epoch := c.fragmentBuffer.pop(); out != nil; out, epoch = c.fragmentBuffer.pop() {
-			rawHandshake := &handshake.Handshake{}
+			var keyExchangeAlgorithm CipherSuiteKeyExchangeAlgorithm
+			if c.state.cipherSuite != nil {
+				keyExchangeAlgorithm = c.state.cipherSuite.KeyExchangeAlgorithm()
+			}
+			rawHandshake := &handshake.Handshake{
+				KeyExchangeAlgorithm: keyExchangeAlgorithm,
+			}
 			if err := rawHandshake.Unmarshal(out); err != nil {
 				c.log.Debugf("%s: handshake parse failed: %s", srvCliStr(c.state.isClient), err)
 				continue
 			}
-
+			if m, ok := rawHandshake.Message.(*handshake.MessageServerHello); ok && m.CipherSuiteID != nil {
+				c.state.cipherSuite = cipherSuiteForID(ciphersuite.ID(*m.CipherSuiteID), c.fsm.cfg.customCipherSuites)
```

Comment:
in `flight3handler` we have a lot more logic and use `findMatchingCipherSuite`. 

I think we need to refactor it so decoding of messages happens when pulling from the cache?

@jkralik what do you think? I am happy to look into it also, this PR is so close it is a frustrating one :)

Reply:
Oh nice I see a way out of this! We shouldn't unmarshal until CipherSuite has been set, let me see if I can get this working.

Will ping you @jkralik when done and get your take. If you are happy it is an approve from me.